File: scala/problem9.scala
/*
https://projecteuler.net/problem=9

Special Pythagorean triplet
*/

object problem9 {
  def baselineAnswer: Long = {
    val product =
      (for {
        a <- 1 until 1000
        b <- a + 1 until 1000
        c = 1000 - a - b
        if c > b
        if (a * a) + (b * b) == (c * c)
      } yield a.toLong * b.toLong * c.toLong).head

    product
  }

  def optimizedAnswer: Long = {
    var a = 1
    while (a < 333) {
      val numerator = 1000L * (500L - a.toLong)
      val denominator = 1000L - a.toLong
      if (denominator != 0 && numerator % denominator == 0L) {
        val b = (numerator / denominator).toInt
        val c = 1000 - a - b
        if (b > a && c > b && (a * a) + (b * b) == (c * c)) {
          return a.toLong * b.toLong * c.toLong
        }
      }
      a += 1
    }
    0L
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 9")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem9.cpp
/*
https://projecteuler.net/problem=9

Special Pythagorean triplet

A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
a^2 + b^2 = c^2.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.
*/

#include <chrono>
#include <iostream>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

long long evaluate_baseline() {
    const int target_sum = 1000;

    for (int a = 1; a < target_sum; ++a) {
        for (int b = a + 1; b < target_sum; ++b) {
            for (int c = b + 1; c < target_sum; ++c) {
                if (a + b + c != target_sum) {
                    continue;
                }
                if (a * a + b * b == c * c) {
                    return static_cast<long long>(a) * b * c;
                }
            }
        }
    }

    return -1;
}

long long evaluate_optimized() {
    const int target_sum = 1000;

    for (int a = 1; a < target_sum / 3; ++a) {
        const int numerator = target_sum * (target_sum - 2 * a);
        const int denominator = 2 * (target_sum - a);

        if (denominator == 0 || numerator % denominator != 0) {
            continue;
        }

        const int b = numerator / denominator;
        const int c = target_sum - a - b;

        if (a < b && b < c && a * a + b * b == c * c) {
            return static_cast<long long>(a) * b * c;
        }
    }

    return -1;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 9" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
